packages/presence-signal/src/createPresenceSignal.ts

```diff
@@ -0,0 +1,108 @@
+import { createSignal, createEffect, onCleanup, type Accessor } from "solid-js";
+
+type SharedTransitionConfig = {
+  /** Duration in milliseconds used both for enter and exit transitions. */
+  transitionDuration: number;
+};
+
+type SeparateTransitionConfig = {
+  /** Duration in milliseconds used for enter transitions (overrides `transitionDuration` if provided). */
+  enterTransitionDuration: number;
+  /** Duration in milliseconds used for exit transitions (overrides `transitionDuration` if provided). */
+  exitTransitionDuration: number;
+};
+
+type Options = (
+  | SharedTransitionConfig
+  | SeparateTransitionConfig
+  | (SharedTransitionConfig & SeparateTransitionConfig)
+) & {
+  /** Opt-in to animating the entering of an element if `isVisible` is `true` during the initial mount. */
+  initialEnter?: boolean;
+};
+
+/**
+ * Animates the appearance of its children.
+ */
+export default function createPresenceSignal(
+  /** Indicates whether the component that the resulting values will be used upon should be visible to the user. */
+  isVisible: Accessor<boolean>,
+  optsValue: Options | Accessor<Options>,
+) {
+  const opts = () => (typeof optsValue === "function" ? optsValue() : optsValue);
+  const exitTransitionDuration = () => {
+    const optsValue = opts();
+    return "exitTransitionDuration" in optsValue
+      ? optsValue.exitTransitionDuration
+      : optsValue.transitionDuration;
+  };
+  const enterTransitionDuration = () => {
+    const optsValue = opts();
+    return "enterTransitionDuration" in optsValue
+      ? optsValue.enterTransitionDuration
+      : optsValue.transitionDuration;
+  };
+
+  const initialEnter = () => opts().initialEnter ?? false;
+  const [animateIsVisible, setAnimateIsVisible] = createSignal(
+    initialEnter() ? false : isVisible(),
+  );
+  const [isMounted, setIsMounted] = createSignal(isVisible());
+  const [hasEntered, setHasEntered] = createSignal(initialEnter() ? false : isVisible());
+
+  const isExiting = () => isMounted() && !isVisible();
+  const isEntering = () => isVisible() && !hasEntered();
+  const isAnimating = () => isEntering() || isExiting();
+
+  createEffect(() => {
+    if (isVisible()) {
+      // `animateVisible` needs to be set to `true` in a second step, as
+      // when both flags would be flipped at the same time, there would
+      // be no transition. See the second effect below.
+      setIsMounted(true);
+    } else {
+      setHasEntered(false);
+      setAnimateIsVisible(false);
+
+      const timeoutId = setTimeout(() => {
+        setIsMounted(false);
+      }, exitTransitionDuration());
+
+      onCleanup(() => clearTimeout(timeoutId));
+    }
+  });
+
+  createEffect(() => {
+    if (isVisible() && isMounted() && !animateIsVisible()) {
+      // Force a reflow so the initial styles are flushed to the DOM
+      if (typeof document !== undefined) {
```

Comment:
Why do we check for document? the code inside `createEffect` won't ever run on the server

Reply:
I was thinking about that when I was converting this because I had the same thought in regards to the original library which had this logic in a `React.useEffect` which has similar constraints. What I realized is that React can be running in non-DOM clients, e.g. `react-native`. This is the only logic in this library that interacts with the DOM directly, so I like the idea of keeping the check. I notice that Solid itself even keeps it's client `render` function in `solid/web` which implies to me that Solid is built in such a way that there is the _potential_ of a `solid-native` concept in the future. Do you think that's enough of a reason to keep the check around ❓ 

It's just a thought, and not a hill I feel the need to die on.

Reply:
I'll remove the document validation from this and leave the reflow check. I think in the future we could consider removing it or not as this moves to the state machine model. That model, I think, would be easier for the user to introduce a reflow if they need it (though I lean towards keeping it generally, just because I don't want users to ever have to really consider it).